src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -240,53 +251,60 @@ private async Task<IEnumerable<NuGetFramework>> GetTargetFrameworksAsync(Package
             return ret;
         }
 
-        private async Task<PackageArchiveReader?> GetPackageArchiveAsync(NuGetReference packageReference, CancellationToken token)
+        private Stream? GetCachedPackage(NuGetReference packageReference)
         {
-            if (packageReference is null)
-            {
-                throw new ArgumentNullException(nameof(packageReference));
-            }
-
-            if (packageReference.GetNuGetVersion() == null)
-            {
-                throw new ArgumentException("Package references must have specific versions to get package archives", nameof(packageReference));
-            }
-
             // First look in the local NuGet cache for the archive
             if (_options.CachePath is string cachePath)
             {
                 var archivePath = Path.Combine(cachePath, packageReference.Name, packageReference.Version, $"{packageReference.Name}.{packageReference.Version}.nupkg");
                 if (File.Exists(archivePath))
                 {
                     _logger.LogDebug("NuGet package {NuGetPackage} loaded from {PackagePath}", packageReference, archivePath);
-                    return new PackageArchiveReader(File.Open(archivePath, FileMode.Open), false);
+                    return File.Open(archivePath, FileMode.Open);
                 }
                 else
                 {
                     _logger.LogDebug("NuGet package {NuGetPackage} not found in package cache", packageReference);
                 }
             }
 
+            return null;
+        }
+
+        private async Task<Stream> GetPackageStream(NuGetReference packageReference, string source, CancellationToken token)
+        {
+            var ms = new MemoryStream();
+            var packageSource = new[] { new PackageSource(source) };
+
+            await DownloadPackageToStreamAsync(packageReference, ms, packageSource, token).ConfigureAwait(false);
+
+            ms.Position = 0;
+
+            return ms;
```

Comment:
Where does this the memorystream object get disposed?

Reply:
The receiver takes ownership. I wish we had a good way of defining that in code, but we don't :(